### PR TITLE
[Comgr] Skip embedded libc++ when system headers exist

### LINKED ISSUE #2445: [Comgr] Embedded libc++ headers conflict with system libstdc++ on RHEL/manylinux
## Summary

Comgr's `COMPILE_SOURCE_WITH_DEVICE_LIBS_TO_BC` action injects embedded libc++ headers at `-idirafter` (lowest priority), assuming any system libstdc++ that's present will work on its own and the embedded headers are only a fallback. On RHEL / manylinux (and any environment with gcc-toolset's libstdc++ installed at clang's default search path), this assumption breaks for HIP `--offload-device-only` compilation.

## Symptom

Any HIP source compiled via the comgr API that includes a C++ stdlib header (e.g. `<tuple>`, `<exception>`, `<stdexcept>`, `<array>`) fails with:

```
include/c++/v1/stddef.h:39:15: fatal error: 'stddef.h' file not found
   39 | #include_next <stddef.h>
```

Reproduced concretely in `compile_hip_with_libcxx_test` running inside `ghcr.io/rocm/therock_build_manylinux_x86_64`. PR #2444 fixes the test by adding `-nostdinc++`, but the underlying behavior still affects external users of the comgr API who are unaware they need that flag.

## Root cause chain

1. Source `#include <tuple>`.
2. Clang resolves `<tuple>` to **system** gcc libstdc++ (e.g. `/usr/lib/gcc/x86_64-redhat-linux/8/.../include/c++/8/tuple`) because the embedded libc++ is injected at `-idirafter`.
3. gcc's `<tuple>` -> `<array>` -> `<stdexcept>` -> `<exception>` -> `<bits/exception_ptr.h>` -> `<bits/cxxabi_init_exception.h>` -> `#include <stddef.h>`.
4. That `<stddef.h>` resolves to **libc++'s** `include/c++/v1/stddef.h` (VFS-mapped to clang's resource dir).
5. libc++'s `stddef.h` does `#include_next <stddef.h>` -- under HIP `--offload-device-only` with `-nogpuinc`, no next stddef.h exists on the device include path.

## Where it lives

`amd/comgr/src/comgr-compiler.cpp:1204-1216`:

```cpp
// Auto-inject embedded libc++ headers as a fallback include path.
// Using -idirafter places them AFTER all other include paths, so:
//   - System libstdc++ or libc++ headers take priority when available
//   - User-provided -I paths take priority
//   - Embedded headers only kick in when no other C++ headers are found
// This ensures backward compatibility while providing headers on systems
// without C++ development headers (e.g., driver-only installs).
if (HasEmbeddedHeaders && getLanguage() == AMD_COMGR_LANGUAGE_HIP) {
  SmallString<256> LibcxxPath(env::getLLVMPath());
  sys::path::append(LibcxxPath, "include", "c++", "v1");
  Argv.push_back("-idirafter");
  Argv.push_back(Saver.save(StringRef(LibcxxPath)).data());
}
```

The comment explicitly assumes "system libstdc++ takes priority when available" is the right default. That assumption is wrong for HIP device-only mode on systems where system libstdc++ headers transitively include `<stddef.h>` and the chain hits a libc++ stddef that wants `#include_next`.

## Possible fixes

1. **Auto-inject `-nostdinc++` when `-nogpuinc` is set and embedded headers are active.** Targeted: only HIP device-only mode would change behavior. Users who want system libstdc++ would have to explicitly drop `-nogpuinc` or pass their own `-I` paths.

2. **Change `-idirafter` -> `-isystem`** for embedded libc++. Embedded headers always take priority. Risk: users who actually depend on system libstdc++ (and don't use `-nogpuinc`) would silently switch to embedded.

3. **Document the requirement** -- tell users they must pass `-nostdinc++` themselves when using `-nogpuinc`. Lowest blast radius, also least helpful.

## Reproducer

```bash
docker run --rm -v /path/to/llvm-project:/llvm:ro -v /tmp/repro:/repro -w /repro \
  ghcr.io/rocm/therock_build_manylinux_x86_64:latest bash -c '
    cmake -G Ninja -S /llvm/llvm -B build -DCMAKE_BUILD_TYPE=Release \
      -DLLVM_ENABLE_PROJECTS="clang;lld" \
      -DLLVM_TARGETS_TO_BUILD="AMDGPU;X86;SPIRV"
    ninja -C build llvm-test-depends clang-test-depends
    cmake -G Ninja -S /llvm/amd/device-libs -B build-device-libs \
      -DCMAKE_PREFIX_PATH=/repro/build -DLLVM_DIR=/repro/build/lib/cmake/llvm
    ninja -C build-device-libs
    cmake -G Ninja -S /llvm/amd/comgr -B build-comgr \
      -DCMAKE_PREFIX_PATH="/repro/build;/repro/build-device-libs" \
      -DLLVM_DIR=/repro/build/lib/cmake/llvm \
      -DBUILD_TESTING=ON
    ninja -C build-comgr compile_hip_with_libcxx_test
    cd build-comgr && LD_LIBRARY_PATH=. AMD_COMGR_REDIRECT_LOGS=stderr \
      ./test/compile_hip_with_libcxx_test
  '
```

(`AMD_COMGR_REDIRECT_LOGS=stderr` is essential -- without it, the failure surfaces only as a generic `AMD_COMGR_STATUS_ERROR` with no detail. This is itself worth surfacing in test infrastructure.)

## Related

- PR #2444: test-level workaround (adds `-nostdinc++`).
- ROCm/SPIRV-LLVM-Translator PR #178: the new SPIRV CI workflow that surfaced this.

### PR DESCRIPTION
**Pull request by yxsamliu**
Fixes #2445.

This PR now contains only the embedded libc++ / system C++ header-mixing fix. The resource-dir / clang path consistency change was split out to #2982. The temporary duplicate split PR #2983 was closed as superseded.

Comgr currently adds its embedded libc++ headers with -idirafter as a fallback. On hosts with system libstdc++ or libc++ headers, that can produce a mixed include chain: host C++ headers first, then Comgr's partial embedded libc++ headers later. In the #2445 failure, libstdc++ reaches a VFS-mapped libc++ stddef.h whose #include_next cannot find the expected next stddef.h under -nogpuinc.

This PR detects common system libc++ and libstdc++ header layouts, including --sysroot and --gcc-toolchain redirects. When system C++ headers are present, Comgr skips the embedded -idirafter path instead of mixing the two header sets.

It also adds AMD_COMGR_USE_EMBEDDED_LIBCXX=force/disable overrides and regression coverage for the libstdc++ conflict path, the system-libc++ detection path, and the forced embedded fallback path.

### CODE REVIEW

**Comment by lamb-j** [comment]
Can this be split into two separate PRs for the two issues? Also comments are a bit verbose (ok if needed for the complexity of the code, but otherwise shorter would be nice)

**Comment by lamb-j** [comment]
Let me revive https://github.com/ROCm/llvm-project/pull/1536, I think it gets at your second issue

**Comment by yxsamliu** [comment]
Thanks Jacob, makes sense to split. Plan: PR A will be a small follow-up to #1536 that updates the VFS embed site in `AMDGPUCompiler` ctor (~line 2417) to also use `getClangExecutable()` — right now it still builds the clang path via `getLLVMPath()` + `sys::path::append`, which is the half causing the resource-dir mismatch with the Driver, and that site needs updating anyway once `getLLVMPath()` is gone. PR B will be the libstdc++/libc++ skip logic stacked on PR A, with the comments trimmed down. I'll wait for #1536 to settle before posting either.

**Comment by yxsamliu** [comment]
Updated this PR to reuse the original #2478 branch for the #2445 embedded libc++ / system C++ header-mixing fix only. The resource-dir / clang path consistency change was split out to #2982, and the temporary duplicate split PR #2983 was closed as superseded.

**Comment by yxsamliu** [comment]
Thanks Jacob. I split the original PR into two pieces.

- #2478 is now only the #2445 fix: skip Comgr's embedded libc++ -idirafter path when system C++ headers are present, to avoid mixing system libstdc++ with the partial embedded libc++ headers.
- #2982 has the separate clang/resource-dir path consistency change. I think it is still needed for the no-system-C++-headers case, because then Comgr falls back to the embedded headers and the builtin include path still needs to match clang's resource-dir path.

I also trimmed the implementation comments in #2478.

**Comment by lamb-j** [comment]
Can you switch the tests to be either GTests or LIT tests? Unless CTest is the best way to test this

**Comment by yxsamliu** [comment]
Updated the PR to use test-lit instead of CTest.

The new tests are under amd/comgr/test-lit/ with a helper for Comgr setup and portable SKIPPED handling when system C++ headers are unavailable.

**Comment by chinmaydd** [comment]
Hi Sam, could you rebase this PR so that the CI can run ?

**Comment by yxsamliu** [comment]
Rebased and pushed. CI was retriggered, but ReadTheDocs is still failing.

**Comment by yxsamliu** [comment]
Could you take another look at this PR when you get a chance?

I rebased it onto `amd-staging` so CI can run, kept the tests under `test-lit`, and fixed one edge case where an explicit `--gcc-toolchain` should not fall back to host/system C++ headers when deciding whether to use Comgr's embedded libc++ fallback.

File: amd/comgr/src/comgr-compiler.cpp [hunk 1]
```diff
@@ -1190,6 +1190,153 @@ amd_comgr_status_t AMDGPUCompiler::removeTmpDirs() {
...
+    return true;
+
+  return false;
+}
+
+bool AMDGPUCompiler::shouldSkipEmbeddedHeaders(ArrayRef<const char *> Argv) {
+  if (SkipEmbeddedHeadersCache)
+    return *SkipEmbeddedHeadersCache;
+
+  bool Verbose = env::shouldEmitVerboseLogs();
```
Comment (chinmaydd) [inline amd/comgr/src/comgr-compiler.cpp:1273]: Diagnostics for every file after the first one would not show the verbose comments. I'm torn on whether that's okay. I'd guess for a multi-file HIP compile, its okay if we only log once.

File: amd/comgr/src/comgr-compiler.cpp [hunk 2]
```diff
@@ -1190,6 +1190,153 @@ amd_comgr_status_t AMDGPUCompiler::removeTmpDirs() {
...
+}
+
+static bool probeLibstdcxxHeadersUnder(StringRef Root, std::string *FoundPath) {
+  auto Hit = [&](const Twine &P) {
+    if (FoundPath)
+      *FoundPath = P.str();
+    return true;
+  };
+
+  SmallString<256> CxxRoot(Root);
```
Comment (chinmaydd) [inline amd/comgr/src/comgr-compiler.cpp]: What about cases where clang is configured with `CLANG_DEFAULT_CXX_STDLIB=libc++`?

File: amd/comgr/src/comgr-compiler.cpp [hunk 3]
```diff
@@ -1190,6 +1190,153 @@ amd_comgr_status_t AMDGPUCompiler::removeTmpDirs() {
...
+
+static bool probeLibcxxHeadersUnder(StringRef Root, std::string *FoundPath) {
+  auto Hit = [&](const Twine &P) {
+    if (FoundPath)
+      *FoundPath = P.str();
+    return true;
+  };
+
+  SmallString<256> LibCxx(Root);
+  sys::path::append(LibCxx, "include", "c++", "v1", "__config_site");
```
Comment (chinmaydd) [inline amd/comgr/src/comgr-compiler.cpp]: We seem to be looking for a specific file that determines the libc++ installation. Does this require documentation of supported versions ?

File: amd/comgr/src/comgr-compiler.cpp [hunk 4]
```diff
@@ -1190,6 +1190,153 @@ amd_comgr_status_t AMDGPUCompiler::removeTmpDirs() {
...
+      Value = A.drop_front(StringRef("-stdlib=").size());
+    else if (A == "-stdlib" && I + 1 < Argv.size() && Argv[I + 1])
+      Value = Argv[++I];
+    else
+      continue;
+
+    if (Value == "libc++")
+      Result = CxxStdlib::Libcxx;
+    else if (Value == "libstdc++")
+      Result = CxxStdlib::Libstdcxx;
```
Comment (chinmaydd) [inline amd/comgr/src/comgr-compiler.cpp]: I guess we can move this out of the loop?

File: amd/comgr/src/comgr-compiler.cpp [hunk 5]
```diff
@@ -1190,6 +1190,153 @@ amd_comgr_status_t AMDGPUCompiler::removeTmpDirs() {
...
+                                 std::string *FoundPath) {
+  if (Stdlib == CxxStdlib::Libcxx)
+    return probeLibcxxHeadersUnder(Root, FoundPath);
+  return probeLibstdcxxHeadersUnder(Root, FoundPath);
+}
+
+// Probe common system C++ header locations, including --sysroot and
+// --gcc-toolchain redirects. Unknown layouts keep the previous embedded
+// fallback behavior.
+static bool detectSystemCxxHeadersOnDisk(ArrayRef<const char *> Argv,
```
Comment (chinmaydd) [inline amd/comgr/src/comgr-compiler.cpp]: Could you perhaps explain how this interacts with the VFS?

Comment (chinmaydd) [reply amd/comgr/src/comgr-compiler.cpp]: I'm thinking if the user can provide something like a `--sysroot` that such that their libcxx installation entirely lives there.

This was inspired by an AI-assisted review, so I'm unsure whether this is a likely scenario

File: amd/comgr/test-lit/comgr-sources/compile-hip-cxx-headers.c [hunk 6]
```diff
@@ -0,0 +1,289 @@
...
+#endif
+}
+
+static int hasClangBuiltinHeadersOnDisk(void) {
+#if defined(_WIN32) || defined(_WIN64)
+  return 0;
+#else
+  const char *LLVMPath = getenv("LLVM_PATH");
+  const char *Roots[] = {LLVMPath ? LLVMPath : "", "/opt/rocm/llvm", "/usr",
+                         NULL};
```
Comment (chinmaydd) [inline amd/comgr/test-lit/comgr-sources/compile-hip-cxx-headers.c:45]: ~~Should we extend this to Windows ? I think you can have Visual Studio installed libstdc++~~

My bad. I just checked and it looks like you cannot get it via VS but only MINGW. I dont think thats a case we realistically need to worry about.

Maybe in this case the MSVC STL acts as the Windows-equivalant of libstdc++ ?

Comment (chinmaydd) [reply amd/comgr/test-lit/comgr-sources/compile-hip-cxx-headers.c:45]: I guess more generally ensuring that after setting the right env variable the behavior is similar across Linux and Win

Comment (chinmaydd) [reply amd/comgr/test-lit/comgr-sources/compile-hip-cxx-headers.c:45]: Or we could add a statement saying this ENV VAR is only supported on Linux distros.

File: amd/comgr/test-lit/compile-hip-system-libcxx.hip [hunk 7]
```diff
@@ -0,0 +1,10 @@
+// COM: System libc++ should only suppress embedded libc++ when requested.
+// RUN: compile-hip-cxx-headers --mode=system-libcxx %s | %FileCheck %s
+
+// CHECK: RESULT: {{PASS|SKIPPED}}
```
Comment (chinmaydd) [inline amd/comgr/test-lit/compile-hip-system-libcxx.hip]: Any reason we are making this check broad ? Might be hard to identify if it regresses

File: amd/comgr/test-lit/compile-hip-distroless-libcxx.hip [hunk 8]
```diff
@@ -0,0 +1,20 @@
+// COM: Force Comgr's embedded libc++ fallback path.
+// RUN: compile-hip-cxx-headers --mode=distroless %s | %FileCheck %s
+
+// CHECK: Embedded libc++ headers: active
+// CHECK: RESULT: PASS
```
Comment (chinmaydd) [inline amd/comgr/test-lit/compile-hip-distroless-libcxx.hip:6]: This test works because `COMGR_HAS_LIBCXX_HEADERS=1` by default. Maybe should be documented.

**Comment by chinmaydd** [comment]
Test failures seem genuine.

**Comment by yxsamliu** [comment]
The remaining CI failures look unrelated to this PR. I checked other open amd-staging PRs and the same gfx94X rocPRIM/hipCUB/rocSPARSE failures appear across many unrelated PRs. The Windows build failure also matches the existing runner file-lock pattern.

**Comment by chinmaydd** [comment]
Some of my comments are still unaddressed

File: amd/comgr/src/comgr-compiler.cpp (hunk 1 above)
Comment (yxsamliu) [reply amd/comgr/src/comgr-compiler.cpp:1273]: This is a verbose Comgr log, not a compiler diagnostic. Since the log is for the whole process, users should be able to understand that the decision also applies to later compilations in the same process.

File: amd/comgr/src/comgr-compiler.cpp (hunk 2 above)
Comment (yxsamliu) [reply amd/comgr/src/comgr-compiler.cpp]: Fixed. The detection now also checks `CLANG_DEFAULT_CXX_STDLIB` when no explicit `-stdlib=` is passed, so a clang configured to default to libc++ will probe libc++ headers.

File: amd/comgr/src/comgr-compiler.cpp (hunk 3 above)
Comment (yxsamliu) [reply amd/comgr/src/comgr-compiler.cpp]: Changed this to check `include/c++/v1/cstddef` instead. That is a normal libc++ header, so this should not need separate version documentation.

File: amd/comgr/src/comgr-compiler.cpp (hunk 4 above)
Comment (yxsamliu) [reply amd/comgr/src/comgr-compiler.cpp]: Done. The loop now only records the last `-stdlib` value, and the enum decision is made after the loop.

File: amd/comgr/src/comgr-compiler.cpp (hunk 5 above)
Comment (yxsamliu) [reply amd/comgr/src/comgr-compiler.cpp]: The disk check runs before adding the embedded libc++ VFS include path. If system C++ headers are found, we skip the embedded VFS fallback so they are not mixed.

For `--sysroot`, the check probes the normal C++ header layouts under the sysroot. If libc++ or libstdc++ is found there, embedded libc++ is skipped. Unknown layouts keep the old embedded fallback behavior.

File: amd/comgr/test-lit/comgr-sources/compile-hip-cxx-headers.c (hunk 6 above)
Comment (yxsamliu) [reply amd/comgr/test-lit/comgr-sources/compile-hip-cxx-headers.c:45]: I think MSVC STL support is out of scope for this PR. This change is focused on the Linux-style issue where system libstdc++/libc++ can be mixed with Comgr embedded libc++ fallback. We can handle Windows/MSVC STL similarly in a future PR if a real need comes up.

File: amd/comgr/include/amd_comgr.h.in [hunk 9]
```diff
@@ -148,6 +148,11 @@ extern "C" {
  *   irrespective of whether logging is enabled.
  * - @p AMD_COMGR_EMIT_VERBOSE_LOGS: If this is set, and is not "0", logs will
  *   include additional Comgr-specific informational messages.
+ * - @p AMD_COMGR_USE_EMBEDDED_LIBCXX: Controls Comgr's embedded libc++ header
```
Comment (lamb-j) [inline amd/comgr/include/amd_comgr.h.in]: I forgot these were even documented here. I'd vote to remove these here, and let the README be the authority, so we don't have to keep both places updated.

That can be done as a follow up though, so fine with this for now

File: amd/comgr/test-lit/compile-hip-system-libcxx.hip (hunk 7 above)
Comment (chinmaydd) [reply amd/comgr/test-lit/compile-hip-system-libcxx.hip]: ping on this

File: amd/comgr/src/comgr-compiler.cpp [hunk 10]
```diff
@@ -1190,6 +1190,153 @@ amd_comgr_status_t AMDGPUCompiler::removeTmpDirs() {
...
+  }
+
+  // User explicitly took control of C++ include search; don't second-guess.
+  for (const char *A : Argv) {
+    if (!A)
+      continue;
+    StringRef S(A);
+    if (S == "-nostdinc++" || S == "-nostdinc" || S == "-nostdlibinc")
+      return Decide(true, Twine("user passed ") + S);
+  }
```
Comment (chinmaydd) [inline amd/comgr/src/comgr-compiler.cpp:1299]: How are we determining the precedence between `-nostdinc` and ENV check ? Whatever be the case, it should be documented

**Review by chinmaydd — APPROVED** [review]
LGTM with nits

File: amd/comgr/test-lit/compile-hip-system-libcxx.hip (hunk 7 above)
Comment (yxsamliu) [reply amd/comgr/test-lit/compile-hip-system-libcxx.hip]: Done. The test now requires `system-libcxx` and checks only `PASS`, so it is skipped by lit when system libc++ headers are unavailable.

File: amd/comgr/include/amd_comgr.h.in (hunk 9 above)
Comment (yxsamliu) [reply amd/comgr/include/amd_comgr.h.in]: Removed it from the public header. The README is now the only place documenting this env var.

File: amd/comgr/src/comgr-compiler.cpp (hunk 10 above)
Comment (yxsamliu) [reply amd/comgr/src/comgr-compiler.cpp:1299]: Documented this in the README. The env var has priority: `force` always uses embedded headers, `disable` never uses them, and the default `auto` mode honors `-nostdinc*` before checking for system C++ headers.

**Comment by skganesan008** [comment]
!PSDB

**Review by lamb-j — APPROVED** [review]

**Comment by chinmaydd** [comment]
@yxsamliu, failures are related to HotSwap, so this PR is technically sound. Could we rebase and get this in ?

**Comment by yxsamliu** [comment]
Rebased and pushed. Local `test-lit` passed. CI is running now.

**Comment by yxsamliu** [comment]
The Windows SPIR-V failure looks unrelated to this PR. The failing step is `Untar build trees`, before the PR-head SPIR-V test runs. The later SPIR-V failures also reproduce on the amd-staging baseline.
